[MOD-2426] Android: Add Clustering to Ti.Map

m1ga opened the pull request:
**JIRA**: https://jira.appcelerator.org/browse/MOD-2426

Adding a property called 'clustering' to the MapView to enable/disable automatic clustering of the annotations depending on the zoom level. Default `false`

```javascript
var Map = require('ti.map');

var win = Titanium.UI.createWindow();

var ann = [];

for (var i = 0; i < 10; i++) {

    ann.push(Map.createAnnotation({
        clusterIdentifier: (i%2==0)?'abc':null,
        collisionMode: Map.ANNOTATION_VIEW_COLLISION_MODE_RECTANGLE,
        showAsMarker:true,
        markerGlyphText:i.toString(),
        markerColor:'blue',
        markerGlyphColor:'green',
        title:'Title',
        subtitle: 'subtitle',
        markerTitleVisibility: Map.FEATURE_VISIBILITY_VISIBLE,
        markerSubtitleVisibility: Map.FEATURE_VISIBILITY_HIDDEN,
        markerGlyphImage:'Flag.png',
        markerSelectedGlyphImage:'social-facebook-icon.png',
        annotationDisplayPriority:Map.FEATURE_DISPLAY_PRIORITY_DEFAULT_LOW,
        latitude : Math.random() * 10 + 40,
        longitude : Math.random() * 10,
    }));
}

var mapview = Map.createView({
    annotations:ann,
    bottom : 0,
    right : 0,
    width : Ti.UI.FILL,
    height : Ti.UI.FILL,
    rotateEnabled : true,
    mapType : Map.MUTED_STANDARD_TYPE,
    showsPointsOfInterest : false,
    userLocation : true,
});

mapview.addEventListener('clusterstart', function(e){
    Ti.API.info('clustering started');
    var clusterAnnotation = Map.createAnnotation({
        showAsMarker:true,
        markerText:e.memberAnnotations.length.toString(),
        title:'Cluster Title',
        subtitle: 'Cluster Subtitle',
    });
    mapview.setClusterAnnotation({
        annotation:clusterAnnotation,
        memberAnnotations:e.memberAnnotations
    });
});
win.add(mapview);
win.open();
```


Since the iOS implementation looks a bit different I'm open to changes.

hansemannn commented:
Looking at #212, can you see possible parity we could reuse here? I'm struggling merging two completely different cluster-approaches, which may be limited by the platform.

hansemannn commented:
@m1ga Driving this forward, how about the following approach to match the iOS parity:
- Instead of using the `clustering` property on the map-view, introduce a more fine-granular configuration via the `clusteringIdentifier` property that indicates which of the annotations are participating in clustering
- Fire the `clusterstart` event once clustering starts. Can this be observed on Android?
- Finally, inside that event, set the annotation that represents the cluster and the annotations it includes:
```js
    var clusterAnnotation = Map.createAnnotation({
        title: e.memberAnnotations.length,
    });
    mapview.setClusterAnnotation({
        annotation: clusterAnnotation, // The annotation that is shown for the included annotations
        memberAnnotations: e.memberAnnotations // The annotations that have "clusteringIdentifier"
    });
```

Is that or some of that possible on Android? Parity would be very critical for this feature. Thanks!

hansemannn commented:
Yey, huge improvement! Would it technically be possible on Android to specify annotations and they are organized internally? Basically setting them and triggering a new cluster after that. But in this state, we could already plan it for the next Ti.Map release!

hansemannn commented:
@garymathews Can you review this when you have time after 7.1.1? Might be a cool parity!

hansemannn commented:
@m1ga I think only the docs are missing right?

m1ga commented:
added Android to the relevant parts of the documentation and 7.2.0 as the since-version.
Also removed the `clustering` property that was still in there from the first commit.

hansemannn commented:
@ypbnv Can you do the review? I think @garymathews is pretty busy with Android 8 changes right now.

ypbnv commented:
@hansemannn On it.

ypbnv commented on android/src/ti/map/AnnotationProxy.java:
```diff
@@ -275,8 +289,10 @@ private void handleImage(Object image)
 			if (bitmap != null) {
-				markerOptions.icon(BitmapDescriptorFactory.fromBitmap(bitmap));
-				setIconImageDimensions(bitmap.getWidth(), bitmap.getHeight());
+				try {
```
Not directly related to this PR, but we could take advantage of `TiDrawableReference.fromObject` in the handling of the image source to streamline the method a bit. Something in the lines of :
```

// this will go through different image sources
TiDrawableReference imageRef = TiDrawableReference.fromObject(this, (String) image);
// check if we loaded the image succesfully
if (!(imageRef.isTypeNull())) {
    // get the bitmap
    Bitmap bitmap = imageref.getBitmap();
    // guard for the bitmap
    if (bitmap != null) {
        try {
            // use it     
            markerOptions.icon(BitmapDescriptorFactory.fromBitmap(bitmap));
            setIconImageDimensions(bitmap.getWidth(), bitmap.getHeight());
        catch (Exception e) {
            e.printStackTrace();
        }
        return;
    }
}
```

m1ga replied:
I left this open, could be handled after the commit

ypbnv commented on android/src/ti/map/TiClusterMarker.java:
```diff
@@ -0,0 +1,52 @@
+
+public class TiClusterMarker implements ClusterItem {
+
+    private LatLng position;
```
Ideally indentation for the file would be with tabs to match the formatting we use for Titanium Java source.

hansemannn replied:
We could use clang-format here, yep! Simply a copy of the .clang-format file and let'd go. Probably after the initial review is approved, so the diff is easier to read.

m1ga replied:
sorry, already did this after making the changes. Since it were mostly some left-overs and intent stuff

ypbnv commented on android/src/ti/map/TiClusterRenderer.java:
```diff
@@ -0,0 +1,75 @@
+        AnnotationProxy anno = clusterItem.getProxy();
+
+        if (anno.hasProperty(TiC.PROPERTY_IMAGE)) {
+            TiDrawableReference imageref = TiDrawableReference.fromUrl(anno, (String) anno.getProperty(TiC.PROPERTY_IMAGE));
```
Maybe we can use the same image handling as in `AnnotationProxy.handleImage()` to support multiple sources?

m1ga replied:
done, it works now with images and blobs

ypbnv commented on android/src/ti/map/TiClusterRenderer.java:
```diff
@@ -0,0 +1,75 @@
+
+public class TiClusterRenderer extends DefaultClusterRenderer<TiClusterMarker> {
+
+    private static final String defaultIconImageHeight = "40dip"; //The height of the default marker icon
```
Indentation with tabs. See above.

m1ga replied:
done

ypbnv commented on android/src/ti/map/TiUIMapView.java:
```diff
@@ -9,6 +9,7 @@
 
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.Collection;
```
Left over import ?

m1ga replied:
removed it

ypbnv commented on android/src/ti/map/TiUIMapView.java:
```diff
@@ -51,11 +52,15 @@
 import com.google.android.gms.maps.model.Marker;
 import com.google.android.gms.maps.model.MapStyleOptions;
+import com.google.maps.android.clustering.ClusterManager;
+import com.google.maps.android.MarkerManager;
```
Another left over?

m1ga replied:
removed it

ypbnv commented on android/src/ti/map/TiUIMapView.java:
```diff
@@ -1092,4 +1145,19 @@ protected void onViewCreated() {
+		if (map != null) {
+			try {
+				map.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, 100));
+			} catch (Exception e) {}
```
I think adding `e.printStackTrace()` could be useful for debugging potential issues around this code.

m1ga replied:
done

m1ga commented:
@hansemannn can do the changes next week

m1ga commented on android/src/ti/map/AnnotationProxy.java:
```diff
@@ -49,7 +49,8 @@
 	MapModule.PROPERTY_CENTER_OFFSET,
-	MapModule.PROPERTY_HIDDEN
+	MapModule.PROPERTY_HIDDEN,
+	MapModule.PROPERTY_CLUSTERIDENTIFIER
```
done
